[HttpFoundation] Adds getAcceptableFormats() method for Request

AndreiIgna opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21909
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features -->

Adds a new method `getAcceptableFormats()` for `Request`. This reads the content types in Accept header and maps them to formats already defined in `Request` class

In a request made by a browser, based on the default Accept header, the `getAcceptableFormats()` will return an array `['html', 'xml']`

In progress
- [x] gather feedback for my changes
- [x] submit changes to the documentation

Commit:
[HttpFoundation] Adds getAcceptableFormats() method for Request

derrabus commented:
It looks like this PR contains unrelated commits. Could you do a rebase against master?

AndreiIgna commented:
Oops used GitHub desktop, which did something unexpected to the rebase. Should be ok now

derrabus commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -170,6 +170,11 @@ class Request
     protected $format;
 
+    /**
+     * @var string
```
Should be `array`.

derrabus commented on src/Symfony/Component/HttpFoundation/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ CHANGELOG
  * The `get()` method of the `AcceptHeader` class now takes into account the
    `*` and `*/*` default values (if they are present in the Accept HTTP header)
    when looking for items.
+ * Adds `getAcceptableFormats()` for reading acceptable formats based on Accept header
```
Should be `added` to be consistent with the wording of other entries in that file.

AndreiIgna commented:
Thanks noticing and letting me know about those @derrabus 

The added method makes it easier to check what data formats the clients accepts, ex:
```
if (in_array('html', $request->getAcceptableFormats())) {
  // return html data, maybe a template
} elseif (in_array('xml', $request->getAcceptableFormats())) {
  // return xml
} else {
  // return json data
}
```

If this is useful and included in next release, I'll make the addition in documentation repo as well

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1371,6 +1376,20 @@ public function getContentType()
+            return $this->acceptableFormats;
+        }
+
+        return $this->acceptableFormats = array_values(array_unique(array_filter(array_map(array($this, 'getFormat'), $this->getAcceptableContentTypes()))));
```
is `array_values` actually required?

AndreiIgna replied:
Is added because the mix of `array_filter` and `array_unqiue` sometimes returns `[0 => 'html', 2 => 'xml', 5 => 'json']`.
This create problems with tests or confusion when checking the array, so it's best to reset the keys

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -1371,6 +1376,20 @@ public function getContentType()
+     *
+     * @return array List of acceptable formats by the client
+     */
+    public function getAcceptableFormats()
```
missing return type hint (and removal of the corresponding docblock line since it will become useless)

nicolas-grekas commented:
@AndreiIgna would you mind rebasing and moving the line in the changelog on 4.2?

AndreiIgna commented:
Yes sure, is updated now

AndreiIgna commented:
Updated based on last review. 

While checking in more detail the similar `$acceptableContentTypes`, saw that it's assigned null on `initialize` and `duplicate` methods. Added this for `$acceptableFormats` too, hope this is ok

nicolas-grekas commented:
Thank you @AndreiIgna.

Tobion commented on src/Symfony/Component/HttpFoundation/Tests/RequestTest.php:
```diff
@@ -1427,6 +1427,16 @@ public function testGetAcceptableContentTypes()
+
+        $request = new Request();
+        $request->headers->set('Accept', 'text/html, application/xhtml+xml, application/xml;q=0.9, */*');
+        $this->assertEquals(array('html', 'xml'), $request->getAcceptableFormats());
```
why would it only return html and xml if everything is acceptable according to `*/*`? I think this shows that the method is not clear and generic enough to be added to the core IMO.

Tobion replied:
I can't see where I would possibly use this method. What would be useful is a similar method that accepts formats as argument and would return the ones that are acceptable according to the accept header. This would then allow to make use of `*/*` and this is what people usually need in REST APIs etc.

nicolas-grekas commented:
FYI, reverted in #29047
